lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -633,6 +701,33 @@ defmodule Trento.Domain.SapSystem do
     end
   end
 
+  defp maybe_emit_sap_system_deregistration_events(
+         %SapSystem{
+           sap_system_id: sap_system_id,
+           application: %Application{
+             instances: instances
+           }
+         },
+         deregistered_at
+       ) do
+    has_abap? = Enum.any?(instances, fn instance -> instance.features =~ "ABAP" end)
+
+    has_messageserver? =
+      Enum.any?(instances, fn instance -> instance.features =~ "MESSAGESERVER" end)
+
+    unless has_abap? and has_messageserver? do
+      Enum.map(instances, fn %Instance{instance_number: instance_number, host_id: host_id} ->
+        %ApplicationInstanceDeregistered{
+          sap_system_id: sap_system_id,
+          instance_number: instance_number,
+          host_id: host_id,
+          deregistered_at: deregistered_at
+        }
+      end) ++
+        [%SapSystemDeregistered{sap_system_id: sap_system_id, deregistered_at: deregistered_at}]
```

Comment:
Hey @rtorrero 
I have noticed that this event should be conditional.
We cannot send this event always as the sap system might not even be registered.
Example:
- We register only message server
- We deregister this same host
At this point, we shouldn't be sending the sap system deregistered event

Reply:
I've added a test for this particular scenario and fixed the code to handle this situation properly. Thanks!